registry/txt.go

```diff
@@ -17,143 +17,127 @@ limitations under the License.
+	result := []*endpoint.Endpoint{}
 
-// ApplyChanges updates dns provider with the changes
-// for each created/deleted record it will also take into account TXT records for creation/deletion
-func (im *TXTRegistry) ApplyChanges(changes *plan.Changes) error {
-	filteredChanges := &plan.Changes{
-		Create:    changes.Create,
-		UpdateNew: filterOwnedRecords(im.ownerID, changes.UpdateNew),
-		UpdateOld: filterOwnedRecords(im.ownerID, changes.UpdateOld),
-		Delete:    filterOwnedRecords(im.ownerID, changes.Delete),
-	}
+	// Go through all non-TXT records and attach the previously collected label sets.
+	for _, r := range records {
+		// Skip all TXT records.
+		if r.RecordType == txtRecordType {
+			continue
+		}
 
-	for _, r := range filteredChanges.Create {
-		txt := endpoint.NewEndpoint(im.mapper.toTXTName(r.DNSName), im.getTXTLabel(), "TXT")
-		filteredChanges.Create = append(filteredChanges.Create, txt)
-	}
-	for _, r := range filteredChanges.Delete {
-		txt := endpoint.NewEndpoint(im.mapper.toTXTName(r.DNSName), im.getTXTLabel(), "TXT")
-		filteredChanges.Delete = append(filteredChanges.Delete, txt)
+		// Attach the label set and add the record to the final list.
+		r.Labels = labels[r.DNSName]
+		result = append(result, r)
 	}
-	return im.provider.ApplyChanges(filteredChanges)
-}
-
-/**
-  TXT registry specific private methods
-*/
-
-func (im *TXTRegistry) getTXTLabel() string {
-	return fmt.Sprintf(txtLabelFormat, im.ownerID)
-}
 
-func (im *TXTRegistry) extractOwnerID(txtLabel string) string {
-	if matches := txtLabelRegex.FindStringSubmatch(txtLabel); len(matches) == 2 {
-		return matches[1]
-	}
-	return ""
+	// Return the endpoints with their labels attached.
+	return result, nil
 }
 
-/**
-  nameMapper defines interface which maps the dns name defined for the source
-  to the dns name which TXT record will be created with
-*/
-
-type nameMapper interface {
-	toEndpointName(string) string
-	toTXTName(string) string
-}
+// ApplyChanges creates, updates and removes a co-located TXT record for each desired endpoint
+// holding information about its labels.
+func (p *TXTRegistry) ApplyChanges(changes *plan.Changes) error {
+	for _, ep := range changes.Create {
+		// Don't create a TXT record if we don't have any labels to remember.
+		if len(ep.Labels) == 0 {
+			continue
+		}
 
-type prefixNameMapper struct {
-	prefix string
-}
+		// For each desired label, prefix the key with ExternalDNS' namespace.
+		labelMap := map[string]string{}
+		for key, value := range ep.Labels {
+			labelMap[labelKeyPrefix+key] = value
+		}
 
-var _ nameMapper = prefixNameMapper{}
+		// Append the TXT record to the original creation list. We also add a special label called
+		// heritage to indicate that this TXT record belongs to ExternalDNS.
+		changes.Create = append(changes.Create, &endpoint.Endpoint{
+			DNSName:    ep.DNSName,
+			Target:     fmt.Sprintf("%s=%s,%s", heritageLabel, heritageValue, formatLabels(labelMap)),
```

Comment:
The heritage label should be added to the `map` and then `formatLabels` should be called. Currently we require to have the heritage label defined first, that's why it's like this atm.